File: src/main/java/game/Tester.java
package game;

import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.Set;

import game.lottery.Lottery;
import game.lottery.LotteryGame;
import game.lottery.LotteryRecord;

public abstract class Tester {

	protected final Lottery lottery;
	protected final int lotterySize;
	protected final int lotteryMin;
	protected final int lotteryMax;
	protected final int numBalls;

	protected List<LotteryRecord> records;
	protected int recordsSize;
	protected int numGamesTotal;
	protected NumberStatistic[][] numStats;

	/**
	 * Creates a new tester.
	 */
	public Tester(Lottery lottery) {
		super();
		this.lottery = lottery;
		this.lotterySize = lottery.size();
		this.lotteryMin = lottery.getMinimum();
		this.lotteryMax = lottery.getMaximum();
		this.numBalls = lottery.getNumberBalls();
	}

	public abstract void parse(File file) throws IOException;

	public abstract void drive();

	protected abstract void drive(NumberStatisticGrouping grouping, String name);

	protected Set<LotteryGame> play(int numGames) {
		Set<LotteryGame> games = lottery.play(numGames);
		int gamesSize = games.size();
		int retry = 10;
		while ((gamesSize < numGames) && (gamesSize > 1)) {
			games.addAll(lottery.play(numGames - gamesSize));
			gamesSize = games.size();
			retry--;
			if (retry == 0) {
				break;
			}
		}
		return games;
	}
}

File: src/main/java/com/github/pnemonic/game/Tester.kt
package com.github.pnemonic.game

import com.github.pnemonic.game.lottery.Lottery
import com.github.pnemonic.game.lottery.LotteryGame
import com.github.pnemonic.game.lottery.LotteryRecord
import java.io.File
import java.io.IOException

abstract class Tester(protected val lottery: Lottery) {
    protected val lotterySize: Int = lottery.size

    protected val lotteryMin: Int = lottery.minimum

    protected val lotteryMax: Int = lottery.maximum

    protected val numBalls: Int = lottery.numberBalls

    protected var records: List<LotteryRecord> = emptyList()

    protected var recordsSize = 0

    protected var numGamesTotal = 0

    protected var numStats: Array<Array<NumberStatistic?>>? = null

    @Throws(IOException::class)
    abstract fun parse(file: File)

    abstract fun drive()

    protected abstract fun drive(grouping: NumberStatisticGrouping, name: String)

    protected fun play(numGames: Int): Set<LotteryGame> {
        val games = lottery.play(numGames)
        var gamesSize = games.size
        var retry = 10
        while (gamesSize in 2 until numGames) {
            games.addAll(lottery.play(numGames - gamesSize))
            gamesSize = games.size
            retry--
            if (retry == 0) {
                break
            }
        }
        return games
    }
}